servers/rendering/shader_preprocessor.cpp

```diff
@@ -1450,7 +1434,27 @@ void ShaderPreprocessor::get_pragma_list(List<String> *r_pragmas) {
 }
 
 ShaderPreprocessor::ShaderPreprocessor() {
+	if (instance_counter == 0) {
+		const String rendering_method = OS::get_singleton()->get_current_rendering_method();
+
+		if (rendering_method == "forward_plus") {
+			defines["CURRENT_RENDERER"] = _MKSTR(2);
+		} else if (rendering_method == "mobile") {
+			defines["CURRENT_RENDERER"] = _MKSTR(1);
+		} else { // gl_compatibility
+			defines["CURRENT_RENDERER"] = _MKSTR(0);
+		}
+
+		defines["RENDERER_COMPATIBILITY"] = _MKSTR(0);
+		defines["RENDERER_MOBILE"] = _MKSTR(1);
+		defines["RENDERER_FORWARD_PLUS"] = _MKSTR(2);
+	}
+	instance_counter++;
 }
 
 ShaderPreprocessor::~ShaderPreprocessor() {
+	instance_counter--;
```

Comment:
Any use of `instance_counter` is subject to possible race conditions, so its not guaranteed to ever reach `0`. If you want to dynamically allocate the defines using a static variable, then you need to either make `instance_counter` atomic, or use a mutex to make this thread safe. 

Alternatively, you can just use a different data structure for "defines" so you aren't dynamically allocating anything.

Reply:
I've changed it to `SafeNumeric`(which internally uses std::atomic), is this fine now? Also I've replace this counter in shader_language.